examples/pipeline_wavernn/main.py

```diff
@@ -0,0 +1,411 @@
+    )
+    parser.add_argument(
+        "--n-fc", default=512, type=int, help="the dimension of fully connected layer",
+    )
+    parser.add_argument(
+        "--kernel-size",
+        default=5,
+        type=int,
+        help="the number of kernel size in the first Conv1d layer",
+    )
+    parser.add_argument(
+        "--n-freq", default=80, type=int, help="the number of bins in a spectrogram",
+    )
+    parser.add_argument(
+        "--n-hidden", default=128, type=int, help="the number of hidden dimensions",
+    )
+    parser.add_argument(
+        "--n-output", default=128, type=int, help="the number of output dimensions",
+    )
+    parser.add_argument(
+        "--n-fft", default=2048, type=int, help="the number of Fourier bins",
+    )
+    parser.add_argument(
+        "--mode",
+        default="waveform",
+        choices=["waveform", "mol"],
+        type=str,
+        help="the mode of waveform",
+    )
+    parser.add_argument(
+        "--seq-len-factor",
+        default=5,
+        type=int,
+        help="seq_length = hop_length * seq_len_factor",
+    )
+    parser.add_argument(
+        "--val-ratio",
+        default=0.1,
+        type=float,
+        help="the ratio of waveforms for validation",
+    )
+    parser.add_argument(
+        "--file-path", default="", type=str, help="the path of audio files",
+    )
+
+    args = parser.parse_args()
+    return args
+
+
+def train_one_epoch(model, mode, criterion, optimizer, data_loader, device, epoch):
+
+    model.train()
+
+    sums = defaultdict(lambda: 0.0)
+    start1 = time()
+
+    metric = MetricLogger("train_iteration")
+    metric["epoch"] = epoch
+
+    for waveform, specgram, target in bg_iterator(data_loader, maxsize=2):
+
+        start2 = time()
+
+        waveform = waveform.to(device)
+        specgram = specgram.to(device)
+        target = target.to(device)
+
+        output = model(waveform, specgram)
+        output, target = output.squeeze(1), target.squeeze(1)
+
+        if mode == "waveform":
+            output = output.transpose(1, 2)
+            target = target.long()
+
+        else:
+            # use mol mode
+            target = target.unsqueeze(-1)
+
+        loss = criterion(output, target)
+        loss_item = loss.item()
+        sums["loss"] += loss_item
+        metric["loss"] = loss_item
+
+        optimizer.zero_grad()
+        loss.backward()
+
+        if args.clip_grad > 0:
+            gradient = torch.nn.utils.clip_grad_norm_(
+                model.parameters(), args.clip_grad
+            )
+            sums["gradient"] += gradient.item()
+            metric["gradient"] = gradient.item()
+
+        optimizer.step()
+
+        metric["iteration"] = sums["iteration"]
+        metric["time"] = time() - start2
+        metric()
+        sums["iteration"] += 1
+
+    avg_loss = sums["loss"] / len(data_loader)
+
+    metric = MetricLogger("train_epoch")
+    metric["epoch"] = epoch
+    metric["loss"] = avg_loss
```

Comment:
nit: you can remove `avg_loss` and replace this by
```
    metric["loss"] = sums["loss"] / len(data_loader)
```

Reply:
... or we can just mimic "evaluate" and leave this as it is.

Reply:
Yes, that is neat. I changed it.